### PR TITLE
Fix race condition in SessionManager.GetSessionInfo

### LINKED ISSUE #14132: Race condition in SessionManager.GetSessionInfo
### Description of the bug

# Description:
The issue happens when the user opens jellyfin, where a websocket connection is initiated and an http request is also initiated from `SessionController` and `PlaystateController`

In this case, the two threads start serving different requests, but both of them tries to get the current session info (mainly via `RequestHelpers.GetSessionId`)

The `GetSessionInfo` method name is misleading, because if the session doesn't exist, it will try to create it. This creation step is not thread safe, and causes the `_activeConnections` to overwrite one session with another. In some races, this causes the session with a websocket controller to get overwritten, causing weird bugs with SyncPlay and other functions that depend on the web sockets..

# Solution 
Using `_activeConnections.GetOrAdd` instead of `_activeConnections.TryGetValue` for an atomic operation.

### Reproduction steps

You have to trigger the race condition by opening the web client for example, on load, the client will initiate a web socket connection and any http request from the above mentioned controllers.

An example stack trace of the code path from the "PostFullCapabilities" controller:

```
Emby.Server.Implementations.Session.SessionManager.GetSessionInfo
Emby.Server.Implementations.Session.SessionManager.LogSessionActivity
Jellyfin.Api.Helpers.RequestHelpers.GetSession
Jellyfin.Api.Helpers.RequestHelpers.GetSessionId
Jellyfin.Api.Controllers.SessionController.PostFullCapabilities
```

An example stack trace of the code path from web socket controller:

```
Emby.Server.Implementations.Session.SessionManager.GetSessionInfo
Emby.Server.Implementations.Session.SessionManager.LogSessionActivity
Emby.Server.Implementations.Session.SessionManager.GetSessionByAuthenticationToken
Emby.Server.Implementations.Session.SessionWebSocketListener.GetSession
Emby.Server.Implementations.Session.SessionWebSocketListener.ProcessWebSocketConnectedAsync
Emby.Server.Implementations.HttpServer.WebSocketManager.WebSocketRequestHandler
```

### What is the current _bug_ behavior?

The sessions created from each "GetSessionInfo" invocations can overwrite each other. 

### What is the expected _correct_ behavior?

A single session should be created, and no multi sessions that can overwrite the other shall exist.

### Jellyfin Server version

10.10.0+

### Specify commit id

_No response_

### Specify unstable release number

_No response_

### Specify version number

_No response_

### Specify the build version

Release

### Environment

```markdown
- OS: Fedora 42
- Linux Kernel: 6.14.6-300.fc42.x86_64
- Virtualization: None
- Clients: Browser
- Browser: 138.0.3 (64-bit)
- FFmpeg Version: ffmpeg version 7.1.1
- Playback Method: Direct Play
- Hardware Acceleration: none
- GPU Model: none
- Plugins: none
- Reverse Proxy: none
- Base URL: none
- Networking: Host
- Storage:local
```

### Jellyfin logs

```shell
N/A as this is a hidden bug with no traces of failure from the system.
```

### FFmpeg logs

```shell

```

### Client / Browser logs

_No response_

### Relevant screenshots or videos

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by moalhaddar**
This PR addresses a thread safety issue in SessionManager.GetSessionInfo where concurrent access could lead to a race condition.

Fixes #14132

### CODE REVIEW

**Review by JPVenson — CHANGES_REQUESTED** [review]
Use GetOrAdd instead

https://learn.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.getoradd?view=net-9.0#system-collections-concurrent-concurrentdictionary-2-getoradd(-0-1)

**Comment by moalhaddar** [comment]
@JPVenson Done.

File: Emby.Server.Implementations/Session/SessionManager.cs [hunk 1]
```diff
@@ -509,13 +509,8 @@ private SessionInfo GetSessionInfo(
 
             var key = GetSessionKey(appName, deviceId);
 
-            CheckDisposed();
```
Comment (JPVenson) [inline Emby.Server.Implementations/Session/SessionManager.cs:512]: why remove the CheckDisposed?

Comment (moalhaddar) [reply Emby.Server.Implementations/Session/SessionManager.cs:512]: It was a duplicate, there are two check disposed being called back to back. The first one is on line 506.

**Review by JPVenson — APPROVED** [review]

**Comment by cvium** [comment]
The change to GetOrAdd now causes OnSessionStarted to sometimes be called. GetOrAdd may run its delegates multiple times

**Comment by moalhaddar** [comment]
After reading the documentation, you're correct. GetOrAdd will run the create session multiple times, creating extra events that might cause unknown side effects.

Is rolling back to the previous lock mechanism preferred in this case?

File: Emby.Server.Implementations/Session/SessionManager.cs [hunk 2]
```diff
@@ -508,13 +510,20 @@ private SessionInfo GetSessionInfo(
...
 
             var key = GetSessionKey(appName, deviceId);
+            SessionInfo sessionInfo;
 
-            CheckDisposed();
-
-            if (!_activeConnections.TryGetValue(key, out var sessionInfo))
+            try
+            {
+                _activeConnectionsLock.EnterWriteLock();
```
Comment (JPVenson) [inline Emby.Server.Implementations/Session/SessionManager.cs:517]: There is no benefit of using a ReaderWriterLock when you dont also lock read operations and only use it to lock write ones

**Review by JPVenson — CHANGES_REQUESTED** [review]

File: Emby.Server.Implementations/Session/SessionManager.cs [hunk 3]
```diff
@@ -508,13 +510,20 @@ private SessionInfo GetSessionInfo(
...
+                if (!_activeConnections.TryGetValue(key, out sessionInfo))
+                {
+                    sessionInfo = CreateSession(key, appName, appVersion, deviceId, deviceName, remoteEndPoint, user);
+                    _activeConnections[key] = sessionInfo;
+                }
+            }
```
Comment (cvium) [inline Emby.Server.Implementations/Session/SessionManager.cs:523]: You are not using the Reader part of the lock. I would suggest this pattern instead with [Lock](https://learn.microsoft.com/en-us/dotnet/api/system.threading.lock?view=net-9.0)

```C#
private readonly Lock _activeConnectionsLock = new();
```

```C#
if (!_activeConnections.TryGetValue(key, out sessionInfo))
{
    using (_activeConnectionsLock.EnterScope())
    {
        if (!_activeConnections.TryGetValue(key, out sessionInfo))
        {
            sessionInfo = CreateSession(key, appName, appVersion, deviceId, deviceName, remoteEndPoint, user);
            _activeConnections[key] = sessionInfo;
        }
    }
}
```

**Comment by moalhaddar** [comment]
@cvium @JPVenson Thank you for the comments. Indeed, the read part if unnecessary. The new code now uses "Lock" instead.

File: Emby.Server.Implementations/Session/SessionManager.cs [hunk 4]
```diff
@@ -508,13 +510,15 @@ private SessionInfo GetSessionInfo(
             ArgumentException.ThrowIfNullOrEmpty(deviceId);
 
             var key = GetSessionKey(appName, deviceId);
+            SessionInfo sessionInfo;
 
-            CheckDisposed();
-
-            if (!_activeConnections.TryGetValue(key, out var sessionInfo))
+            using (_activeConnectionsLock.EnterScope())
```
Comment (cvium) [inline Emby.Server.Implementations/Session/SessionManager.cs:515]: I think GetSessionInfo gets called a lot. This might be a lot of unnecessary locking

Comment (moalhaddar) [reply Emby.Server.Implementations/Session/SessionManager.cs:515]: When i traced the method calls, the GetSessionInfo is only called in this scenarios:
- Creating a new websocket connection (ProcessWebSocketConnectedAsync in SessionWebSocketListener)
- AuthorizeQuickConnect in the QuickConnectController
- AuthenticateUserByName in the UserController

From that, i can see the GetSessionInfo does not fall into a hot path of being repeatedly called from the clients, it's only called once from the above controllers/websocket

Comment (moalhaddar) [reply Emby.Server.Implementations/Session/SessionManager.cs:515]: @cvium  After going through the code again, i believe you're correct that this is a hot path for some controllers. However, this lock is necessary with the current design of the session creation/retrieval.

Let me know if you have other suggestions.

**Comment by gnattu** [comment]
What's my idea on fixing this:

1. Separate the `CreateSession` function with the pure function part and the side effect part. So that we can have one `CreateSessionInfo` function that only returns a new session Info, but don't start this session yet.
2. Then, we can use the `GetOrAdd` to add this created `SessionInfo` to `_activeConnections`.
3. Each thread can do a reference comparison to determine if the object currently in the dictionary is the one current thread created. If so, it means current thread win the race, and we invoke the `OnSessionStarted` callback on current thread with the `SessionInfo` created by current thread.

This approach uses slightly more memory as it may create multiple `SessionInfo` objects on multiple threads but it should perform better than a huge lock for everything.

**Comment by moalhaddar** [comment]
Great idea! @gnattu 
I've went and implemented that change in the PR, I've also tested the change locally with no side effects (other than the minimum memory for the unused sessioninfos).

File: Emby.Server.Implementations/Session/SessionManager.cs [hunk 5]
```diff
@@ -508,13 +508,17 @@ private SessionInfo GetSessionInfo(
...
+            if (ReferenceEquals(newSession, existingOrNewSession))
+            {
+                OnSessionStarted(newSession);
+                sessionInfo = newSession;
+            }
+            else
             {
-                sessionInfo = CreateSession(key, appName, appVersion, deviceId, deviceName, remoteEndPoint, user);
-                _activeConnections[key] = sessionInfo;
+                sessionInfo = existingOrNewSession;
```
Comment (gnattu) [inline Emby.Server.Implementations/Session/SessionManager.cs:521]: This does not really need an else branch. It uses existingOrNewSession in both cases.

**Review by gnattu — APPROVED** [review]

**Comment by moalhaddar** [comment]
@JPVenson  @cvium The new solution does not introduce any locks, please let me know if you have other comments.

**Review by JPVenson — APPROVED** [review]

**Review by crobibero — APPROVED** [review]
